gu-client/src/async.rs

```diff
@@ -35,161 +52,193 @@ impl HubConnection {
+            })
+    }
 }
 
 /// Hub session.
 #[derive(Clone, Debug)]
 pub struct HubSession {
     hub_connection: HubConnection,
-    session_id: String,
+    session_id: u64,
 }
 
 impl HubSession {
     /// adds peers to the hub session
-    pub fn add_peers<T, U>(&self, peers: T) -> impl Future<Item = Vec<NodeId>, Error = Error>
+    pub fn add_peers<T, U: TryIntoNodeId>(
+        &self,
+        peers: T,
+    ) -> impl Future<Item = Vec<NodeId>, Error = Error> + 'static
     where
         T: IntoIterator<Item = U>,
-        U: AsRef<str>,
     {
         let add_url = format!(
             "{}sessions/{}/peers",
-            self.hub_connection.hub_connection_inner.url, self.session_id
+            self.hub_connection.url(),
+            self.session_id
         );
-        let peer_vec: Vec<String> = peers.into_iter().map(|peer| peer.as_ref().into()).collect();
-        let request = match client::ClientRequest::post(add_url).json(peer_vec) {
+
+        let peers = match peers
+            .into_iter()
+            .map(|peer| TryIntoNodeId::into_node_id(peer))
+            .collect::<Result<Vec<_>, _>>()
+        {
+            Ok(p) => p,
+            Err(e) => return future::Either::A(future::err(Error::Other(format!("{}", e)))),
+        };
+
+        let request = match client::ClientRequest::post(add_url).json(peers) {
             Ok(r) => r,
-            Err(e) => return future::Either::A(future::err(Error::CannotCreateRequest(e))),
+            Err(e) => return future::Either::A(future::err(Error::CreateRequest(e))),
         };
-        let session_id_copy = self.session_id.clone();
+        let session_id = self.session_id.clone();
+
         future::Either::B(
             request
                 .send()
-                .map_err(Error::CannotSendRequest)
+                .from_err()
                 .and_then(|response| match response.status() {
                     http::StatusCode::NOT_FOUND => {
-                        future::Either::A(future::err(Error::SessionNotFound(session_id_copy)))
+                        future::Either::A(future::err(Error::ResourceNotFound))
+                    }
+                    http::StatusCode::INTERNAL_SERVER_ERROR => {
+                        future::Either::A(future::err(Error::ResponseErr(response.status())))
                     }
-                    http::StatusCode::INTERNAL_SERVER_ERROR => future::Either::A(future::err(
-                        Error::CannotAddPeersToSession(response.status()),
-                    )),
-                    _ => future::Either::B(
-                        response.json().map_err(|e| Error::InvalidJSONResponse(e)),
-                    ),
+                    _ => future::Either::B(response.json().from_err()),
                 }),
         )
     }
     /// creates a new blob
-    pub fn new_blob(&self) -> impl Future<Item = Blob, Error = Error> {
+    pub fn new_blob(&self) -> impl Future<Item = Blob, Error = Error> + 'static {
```

Comment:
this probably belongs to `Blob` impl. I propose to provide static factory method there with new blob creation

It might be wrapped here for convenience

Reply:
Why ? To create blob you need to create new hub session. 

What is better ? 

a) 
```
connection.new_session(spec).and_then(|hub_session| {
    let blob_fut = hub_session.new_blob();
   
   blob_fut.and_then(|blob| 
     hub_session.update(...DownloadFile { uri: blob.uri(), }))
})
```
b) Yours
```
connection.new_session(spec).and_then(|hub_session| {
    let blob_fut = Blob::new_for(hub_session);
   
   blob_fut.and_then(|blob| 
     hub_session.update(...DownloadFile { uri: blob.uri(), }))
})
```

Reply:
it's about handling `Blob` operations within `Blob` impl. Better encapsulation.
Will not insist. This is not a blocker